Fixed copying of arrays to internal HEAP memory through the WebIDL interface.

ondys opened the pull request:
In case the heap memory (HEAP8, HEAP32, ...) was reallocated inside the ensureCache.alloc() function, the content of the array was still copied to the old HEAP, causing it to be lost/discarded. I split the alloc and copy to two functions to ensure that copy is always performed on the correct HEAP.

kripken commented:
Nice find! Yeah, malloc might reallocate those buffers, so we do need to separate this as you did.

Can you please redirect this PR to the `incoming` branch? Also add yourself to AUTHORS.

It would also be good to test this, but if that's too hard, maybe not worth it. But I think we could, we can set total memory to something fairly small and do a loop with enough allocations that it eventually grows memory. Can perhaps add this to the end of `test_webidl` in `tests/test_core.py`.

ondys commented:
@kripken I've modified the WebIDL unit test to detect this problem in future. I wasn't really sure how to test it directly in `tests/test_core.py` so I added the test to `tests/webidl/post.js` + I had to allow memory reallocation in `tests/test_core.py`. The test fails without the fix that I posted. I'm not really that familiar with your testing environment so please let me know if this is acceptable.

kripken commented:
Thanks, the test looks mostly good, however, let's run it both with and without memory growth (without, it should ignore this part of the test). The reason this is important is that we do need to keep testing the non-memory growth mode, often bugs show up just in one, and it matters more.

kripken commented on tests/webidl/post.js:
```diff
@@ -237,6 +237,32 @@ for (var i = 0; i < 1000000; i++) {
+    }
+  }
+}
+
```
let's also ensure that reallocation actually happened. I think you can just look at `Module['HEAP8'].length` before and after all the work, and see that it has in fact grown.

ondys commented:
Modified the test according to your comments. Memory reallocation is now tested separately only when memory growth is enabled.

kripken commented:
Thanks! Looks good, I see a failure when I run e.g. `./tests/runner.py asm2.test_webidl`, on `ensureInt32`. Looks like the post.js code is added after closure compiler runs, so it minifies the names. We could refactor the test to include the post.js more properly, but instead, we could just call a compiled method that uses that (which would also be more future proof, as ensureInt32 is an internal detail).

ondys commented:
Done. Added a simple class that implicitly calls the `ensureInt32` function. All instances of `test_webidl` pass.

kripken commented:
Thanks!
